### ISSUE TITLE
Opt-in form element styling

### ISSUE THREAD

**Issue by levimcg** [indiana-university/rivet-source#250]
This may be more appropriate to write up as an RFC, but I wanted to create an issue here to possibly get some feedback. 

In Rivet 1.x.x we applied all form element styles *globally* using element selectors not CSS classes. This means that styles are automatically applied to the following elements:

- `input[type="text"]` including all HTML5 variants e.g. `type="search"`
- `select`
- `textarea`
- `input[type="checkbox"]`
- `input[type="radio"]`
- `label`
- `fieldset`
- `legend`

I would like to propose that instead of styling the HTML element selectors directly that we start using CSS classes applied to form elements to handle styling moving forward.

Take the following example of a form with one text input and a checkbox:

```html
<form>
  <label for="email">Email address</label>
  <input type="text">

  <input type="checkbox" id="agree">
  <label for="agree">Agree to terms of service</label>

  <button class="rvt-button" type="submit">Submit</button>
</form>
```

In Rivet 1.x.x, this form element (without CSS classes) would automatically be style to look like Rivet form inputs. Moving to *opt-in* styling for form elements would mean that developers would instead need to explicitly apply CSS classes to form elements to style them.

```html
<form>
  <label class="rvt-label" for="email">Email address</label>
  <input class="rvt-text-input" type="text">
  
  <input class="rvt-checkbox" type="checkbox" id="agree">
  <label class="rvt-label" for="agree">Agree to terms of service</label>
  
  <button class="rvt-button" type="submit">Submit</button>
</form>
```

## Why opt-in styling?

For the most part all other elements and components in Rivet require adding CSS classes to style them. Only form elements and a handful of other global styles are applied directly to HTML element selectors.

### Consistency
Form elements make up the bulk of CSS selectors in Rivet that style the HTML elements directly without explicitly applying CSS classes. Moving toward an opt-in model would make the way form element styling is applied more consistent with the rest of Rivet. 

Take the button element for example. In Rivet you need to explicitly apply the `rvt-button` CSS class to a button to get Rivet button styles. This is really helpful since buttons are often times used inside of other interactive components for JavaScript functionality. Since button styles are *opt-in* when we build a new Rivet component that needs to use a button, but needs to appear differently than a default Rivet button, we don't have to undo the Rivet button styles to do so.

### Sometime developers don't have control of markup
Rivet use custom CSS to improve visual design and accessibility of radios and checkboxes by hiding the default browser-rendered radio/checkbox and using the `::before` pseudo element of the `label` to re-create the a visual representation of the radio/checkbox. This helps improve color contrast and increases the hit target size of radios and checkboxes.

Custom radios and checkboxes are great from a visual design and accessibility perspective but can be tricky when a developer does not have control of the rendered markup, e.g. some application frameworks (.NET, Laravel, etc.) auto generate some form markup as a convenience. Rivet radios and checkboxes rely on the the `input` element to come after the `label` element, so if the rendered markup is not in that order the checkbox or radio will not render because the element itself is hidden (see above) and the `::before` element is styled. If form element styling was opt-in this auto generated markup would not match Rivet's form styles, but the controls would at least be visible and usable.

### Partial Rivet Implementation
 Opt-in styling would also help make it easier for teams that may be implementing Rivet as they are rebuilding parts of legacy sites and applications. Teams in this situation may need to keep some parts of their applications that include forms elements as is. Since Rivet currently uses default HTML element selectors for form elements, this can cause a lot of style collisions between Rivet's CSS and legacy CSS.

## Implementation details
For the most part this should be a pretty low-effort change. For the most part it would mean taking styles applied via generic HTML element selectors and moving them to a new block with a class name selector.

### Example
Take the `select` element for as an example.

```scss
select {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxNiIgaGVpZ2h0PSIxNiIgdmlld0JveD0iMCAwIDE2IDE2Ij4KICAgIDxwYXRoIGZpbGw9IiMzMzMzMzMiIGQ9Ik04LDEyLjQ2YTIsMiwwLDAsMS0xLjUyLS43TDEuMjQsNS42NWExLDEsMCwxLDEsMS41Mi0xLjNMOCwxMC40Nmw1LjI0LTYuMTFhMSwxLDAsMCwxLDEuNTIsMS4zTDkuNTIsMTEuNzZBMiwyLDAsMCwxLDgsMTIuNDZaIi8+Cjwvc3ZnPg==');
  background-position: right $sm center;
  background-size: $sm $sm;
  background-repeat: no-repeat;
  padding-right: $xl;
  padding-left: $xs;

  /**
    * Prefixes get rid of browser default select element styles.
    */
  -webkit-appearance: none;
  -moz-appearance: none;
  -o-appearance: none;
  appearance: none;
}


// This would become

.rvt-select {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxNiIgaGVpZ2h0PSIxNiIgdmlld0JveD0iMCAwIDE2IDE2Ij4KICAgIDxwYXRoIGZpbGw9IiMzMzMzMzMiIGQ9Ik04LDEyLjQ2YTIsMiwwLDAsMS0xLjUyLS43TDEuMjQsNS42NWExLDEsMCwxLDEsMS41Mi0xLjNMOCwxMC40Nmw1LjI0LTYuMTFhMSwxLDAsMCwxLDEuNTIsMS4zTDkuNTIsMTEuNzZBMiwyLDAsMCwxLDgsMTIuNDZaIi8+Cjwvc3ZnPg==');
  background-position: right $sm center;
  background-size: $sm $sm;
  background-repeat: no-repeat;
  padding-right: $xl;
  padding-left: $xs;

  /**
    * Prefixes get rid of browser default select element styles.
    */
  -webkit-appearance: none;
  -moz-appearance: none;
  -o-appearance: none;
  appearance: none;
}
```

And be applied to markup like so:

```html
<label class="rvt-label" for="my-select">Select one option</label>
<select class="rvt-select" id="my-select">
  <option>Option one</option>
  <option>Option two</option>
  <option>Option three</option>
</select>
```

## Other considerations
I think if we decide to go this route we should also look at applying this same convention anywhere else we are using generic HTML element selector (within reason). For example list elements like `ul` , `ol`, and `dl`.

**Comment by levimcg** [comment]
I've shared this around a bit internally to get some feedback, but so far not gotten any engagement.  I'd like to go ahead and move forward with implementing this type of _opt-in_ strategy for form elements. Still happy to hear any feedback, but after thinking on it for a while, I think this is the right way forward.

**Comment by illusivesunrae** [comment]
Released with PR #294.
